README.md
[![Build Status](https://travis-ci.org/nodeshift-starters/nodejs-health-check.svg?branch=master)](https://travis-ci.org/nodeshift-starters/nodejs-health-check) [![Coverage Status](https://coveralls.io/repos/github/nodeshift-starters/nodejs-health-check/badge.svg?branch=master)](https://coveralls.io/github/nodeshift-starters/nodejs-health-check?branch=master) [![Greenkeeper badge](https://badges.greenkeeper.io/nodeshift-starters/nodejs-health-check.svg)](https://greenkeeper.io/)

# DEPRECATED - This is no longer supported please consider using 10.x instead

## Running The Example

You can run this example as node processes on your localhost, as pods on a local
[minishift](https://github.com/minishift/minishift/releases) installation.

### Localhost

To run the basic application on your local machine, just run the commands bellow:

```
$ npm install
$ npm start
```

If you want debug information, you can set `DEBUG` environment variable and start the application:

```
$ DEBUG=* npm start
```

This will launch the application on port 8080.

### Minishift

Minishift should be started, and you should be logged in with a currently
active project. Then run the `npm run openshift` command.

```sh
$ minishift start # You may have some options here, e.g. --memory=8096 --vm-driver=virtualbox
$ oc login -u developer # Login
$ oc new-project my-example-project # Create a project to deploy to
$ npm run openshift # Deploys the example app
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [4.0.0](https://github.com/nodeshift-starters/nodejs-health-check/compare/v3.0.2...v4.0.0) (2020-11-26)


### Features

* upgrade tape from 4.13.2 to 5.0.1 ([#186](https://github.com/nodeshift-starters/nodejs-health-check/issues/186)) ([00d27fc](https://github.com/nodeshift-starters/nodejs-health-check/commit/00d27fc270ce2b3d5d0b9ea18ba38bb4a175954d))


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([#180](https://github.com/nodeshift-starters/nodejs-health-check/issues/180)) ([6b7ce31](https://github.com/nodeshift-starters/nodejs-health-check/commit/6b7ce316afde959bb44df7a00e2c38521d52d7cc))
* package.json & package-lock.json to reduce vulnerabilities ([#182](https://github.com/nodeshift-starters/nodejs-health-check/issues/182)) ([fbdb2d6](https://github.com/nodeshift-starters/nodejs-health-check/commit/fbdb2d6006295d01293d22bf02766a1ae9873d90))
* upgrade coveralls from 3.0.11 to 3.1.0 ([#187](https://github.com/nodeshift-starters/nodejs-health-check/issues/187)) ([c74d018](https://github.com/nodeshift-starters/nodejs-health-check/commit/c74d018a15e0fc118d56868bccdbc29040134a97))
* upgrade coveralls from 3.0.9 to 3.0.11 ([#176](https://github.com/nodeshift-starters/nodejs-health-check/issues/176)) ([c9b8f3c](https://github.com/nodeshift-starters/nodejs-health-check/commit/c9b8f3cdcf85532522f9894fd0b7b72f646feb2e))
* upgrade debug from 4.1.1 to 4.2.0 ([#178](https://github.com/nodeshift-starters/nodejs-health-check/issues/178)) ([13b1d82](https://github.com/nodeshift-starters/nodejs-health-check/commit/13b1d82f6e1e08fd57845b9b2a6dc8e4b896b906))
* upgrade kube-probe from 0.4.0 to 0.5.0 ([#189](https://github.com/nodeshift-starters/nodejs-health-check/issues/189)) ([da26fff](https://github.com/nodeshift-starters/nodejs-health-check/commit/da26fffb24e25a7a5f18e1f70767bc2579ccad6f))
* upgrade nodeshift from 6.0.2 to 6.2.0 ([#188](https://github.com/nodeshift-starters/nodejs-health-check/issues/188)) ([9542ed3](https://github.com/nodeshift-starters/nodejs-health-check/commit/9542ed3064297a4c0a78b5dfa258456f26e5ef33))

### [3.0.2](https://github.com/nodeshift-starters/nodejs-health-check/compare/v3.0.1...v3.0.2) (2020-04-27)

### [3.0.1](https://github.com/nodeshift-starters/nodejs-health-check/compare/v3.0.0...v3.0.1) (2019-11-27)

## [3.0.0](https://github.com/nodeshift-starters/nodejs-health-check/compare/v2.0.3...v3.0.0) (2019-05-31)


### Features

*  Update to use Node.js 12.x ([26ea7f3](https://github.com/nodeshift-starters/nodejs-health-check/commit/26ea7f3))


### BREAKING CHANGES

* start using node 12



## [2.0.3](https://github.com/nodeshift-starters/nodejs-health-check/compare/v2.0.2...v2.0.3) (2019-04-15)



<a name="2.0.2"></a>
## [2.0.2](https://github.com/nodeshift-starters/nodejs-health-check/compare/v2.0.1...v2.0.2) (2019-01-11)



<a name="2.0.1"></a>
## [2.0.1](https://github.com/nodeshift-starters/nodejs-health-check/compare/v2.0.0...v2.0.1) (2018-11-08)


### Bug Fixes

* travis-ci should use npm install instead of npm ci ([a8e8822](https://github.com/nodeshift-starters/nodejs-health-check/commit/a8e8822))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/nodeshift-starters/nodejs-health-check/compare/v1.2.3...v2.0.0) (2018-05-16)


### Features

* start using node.js 10. ([20c2c2b](https://github.com/nodeshift-starters/nodejs-health-check/commit/20c2c2b)), closes [#104](https://github.com/nodeshift-starters/nodejs-health-check/issues/104)



<a name="1.2.3"></a>
## [1.2.3](https://github.com/nodeshift-starters/nodejs-health-check/compare/v1.2.2...v1.2.3) (2018-04-26)


### Bug Fixes

* use the main prop for the application entry point. ([a349997](https://github.com/nodeshift-starters/nodejs-health-check/commit/a349997)), closes [nodeshift-starters/centos7-s2i-nodejs#33](https://github.com/nodeshift-starters/centos7-s2i-nodejs/issues/33) [#issuecomment-382587104](https://github.com/nodeshift-starters/nodejs-health-check/issues/issuecomment-382587104) [#101](https://github.com/nodeshift-starters/nodejs-health-check/issues/101)
* **www:** debug should take the project name as a parameter ([ffedf23](https://github.com/nodeshift-starters/nodejs-health-check/commit/ffedf23))



<a name="1.2.2"></a>
## [1.2.2](https://github.com/nodeshift-starters/nodejs-health-check/compare/v1.2.1...v1.2.2) (2018-04-16)



<a name="1.2.1"></a>
## [1.2.1](https://github.com/nodeshift-starters/nodejs-health-check/compare/v1.2.0...v1.2.1) (2018-03-07)


### Bug Fixes

* **package:** need to run the licenser after the version bump ([80dd106](https://github.com/nodeshift-starters/nodejs-health-check/commit/80dd106))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/nodeshift-starters/nodejs-health-check/compare/v1.1.1...v1.2.0) (2018-03-07)


### Bug Fixes

* response text for not online greeting needed to be updated. ([e47508c](https://github.com/nodeshift-starters/nodejs-health-check/commit/e47508c)), closes [#87](https://github.com/nodeshift-starters/nodejs-health-check/issues/87)
* small typo fix ([5a0af0b](https://github.com/nodeshift-starters/nodejs-health-check/commit/5a0af0b)), closes [#76](https://github.com/nodeshift-starters/nodejs-health-check/issues/76)
* **application.yaml:** docker tag should be 8.x ([0d8210c](https://github.com/nodeshift-starters/nodejs-health-check/commit/0d8210c))


### Features

* **package:** Move nsp check from prepare step to an optional script. ([#86](https://github.com/nodeshift-starters/nodejs-health-check/issues/86)) ([2b61e13](https://github.com/nodeshift-starters/nodejs-health-check/commit/2b61e13)), closes [#85](https://github.com/nodeshift-starters/nodejs-health-check/issues/85)


test/greeting-test.js
'use strict';

const test = require('tape');
const supertest = require('supertest');
const proxyquire = require('proxyquire');
const app = require('../app');

test('test out greeting route with no query param', t => {
  supertest(app)
    .get('/api/greeting')
    .expect('Content-Type', /json/)
    .expect(200)
    .then(response => {
      t.equal(response.body.content, 'Hello, World!');
      t.end();
    });
});

test('test out greeting route with a query param', t => {
  supertest(app)
    .get('/api/greeting?name=Luke')
    .expect('Content-Type', /json/)
    .expect(200)
    .then(response => {
      t.equal(response.body.content, 'Hello, Luke');
      t.end();
    });
});

test('test out greeting route after /stop route', t => {
  supertest(app)
    .get('/api/stop')
    .expect(200)
    .then(response => {
      t.equal(response.text, 'Stopping HTTP server', 'stop endpoint response text');
      return supertest(app)
        .get('/api/greeting')
        .expect(503);
    }).then(response => {
      t.equal(response.text, 'Not online', 'should have a not online text');
      t.end();
    });
});

test('test livenessCallback returns status OK', t => {
  const mockres = {
    send: status => {
      t.equal(status, 'OK');
      t.end();
    }
  };
  const mockProbe = {
    init: (expressApp, options) => {
      this.options = options;
    },
    trigger: () => {
      this.options.livenessCallback(null, mockres);
    }
  };
  const proxyApp = proxyquire('../app', { 'kube-probe': mockProbe.init });
  supertest(proxyApp)
    .get('/api/greeting?name=Luke')
    .expect('Content-Type', /json/)
    .expect(200)
    .then(() => {
      mockProbe.trigger();
    });
});
test('test livenessCallback returns statusCode 500', t => {
  const mockres = {
    sendStatus: statusCode => {
      t.equal(statusCode, 500);
      t.end();
    }
  };

  const mockProbe = {
    init: (expressApp, options) => {
      this.options = options;
    },
    trigger: () => {
      this.options.livenessCallback(null, mockres);
    }
  };
  const proxyApp = proxyquire('../app', { 'kube-probe': mockProbe.init });
  supertest(proxyApp)
    .get('/api/stop')
    .expect(200)
    .then(() => {
      mockProbe.trigger();
    });
});


test/integration/greeting-integration-test.js
'use strict';

const util = require('util');
const test = require('tape');
const supertest = require('supertest');
const rhoaster = require('rhoaster');

const setTimeoutPromise = util.promisify(setTimeout);

const packagejson = require('../../package.json');

const testEnvironment = rhoaster({
  deploymentName: packagejson.name,
  dockerImage: 'registry.access.redhat.com/ubi8/nodejs-12'
});

testEnvironment.deploy()
  .then(runTests)
  .then(_ => test.onFinish(testEnvironment.undeploy))
  .catch(console.error);

function runTests (route) {
  test('/api/greeting', t => {
    t.plan(1);
    supertest(route)
      .get('/api/greeting')
      .expect(200)
      .expect('Content-Type', /json/)
      .then(response => {
        t.equal(response.body.content, 'Hello, World!', 'should return the Hello, World Greeting');
        t.end();
      });
  });

  test('/api/greeting with query param', t => {
    t.plan(1);
    supertest(route)
      .get('/api/greeting?name=luke')
      .expect(200)
      .expect('Content-Type', /json/)
      .then(response => {
        t.equal(response.body.content, 'Hello, luke', 'should return the Hello, luke Greeting');
        t.end();
      });
  });

  test('/api/health/readiness returns OK', async t => {
    // Wait the initial 60 seconds before the liveness probe kicks in.
    await setTimeoutPromise(60000);
    supertest(route)
      .get('/api/health/readiness')
      .expect(200)
      .expect('Content-Type', 'text/html')
      .then(response => {
        t.equal(response.text, 'OK', 'should return OK');
        // Now shut the endpoint down
        return supertest(route)
          .get('/api/stop')
          .expect(200);
      })
      .then(response => {
        t.equal(response.text, 'Stopping HTTP server', 'should be stopping the server');
        // Check that the greeting is down
        return supertest(route)
          .get('/api/greeting')
          .expect(503);
      })
      .then(response => {
        t.equal(response.statusCode, 503, 'shold have a 400 response');
        // Now ping the liveness probe which should return a 50x response
        return supertest(route)
          .get('/api/health/liveness')
          .expect(500);
      })
      .then(response => {
        t.equal(response.statusCode, 500, 'shold have a 500 response');
      })
      .then(_ => {
        // Wait until the app is back up
        return setTimeoutPromise(20000);
      }).then(_ => {
        // After we wait check the app again
        supertest(route)
          .get('/api/greeting')
          .expect(200)
          .expect('Content-Type', /json/)
          .then(response => {
            t.equal(response.body.content, 'Hello, World!', 'should return the Hello, World Greeting');
            t.end();
          });
      });
  });
}


release.js
#!/usr/bin/env node

'use strict';

/*
 *
 *  Copyright 2016-2017 Red Hat, Inc, and individual contributors.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

// This file is run during the "postbump" lifecyle of standard-version
// We need to be able to update the metadata.label.verion of the resource objects in the openshift template in the .openshiftio folder

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const jsyaml = require('js-yaml');
const packagejson = require('./package.json');

const writeFile = promisify(fs.writeFile);
const readFile = promisify(fs.readFile);

async function updateApplicationYaml () {
  const applicationyaml = jsyaml.safeLoad(await readFile(path.join(__dirname, '/.openshiftio/application.yaml'), { encoding: 'utf8' }));

  // We just need to update the RELEASE_VERSION parameter
  applicationyaml.parameters = applicationyaml.parameters.map(param => {
    if (param.name === 'RELEASE_VERSION') {
      param.value = packagejson.version;
    }

    return param;
  });

  // Now write the file back out
  await writeFile(path.join(__dirname, '/.openshiftio/application.yaml'), jsyaml.safeDump(applicationyaml, { skipInvalid: true }), { encoding: 'utf8' });
}

updateApplicationYaml();


app.js
'use strict';

/*
 *
 *  Copyright 2016-2017 Red Hat, Inc, and individual contributors.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

const path = require('path');
const express = require('express');
const bodyParser = require('body-parser');
const probe = require('kube-probe');

const app = express();

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(express.static(path.join(__dirname, 'public')));
// Expose the license.html at http[s]://[host]:[port]/licences/licenses.html
app.use('/licenses', express.static(path.join(__dirname, 'licenses')));

let isOnline = true;

//
app.use('/api/greeting', (request, response) => {
  if (!isOnline) {
    response.status(503);
    return response.send('Not online');
  }

  const name = request.query ? request.query.name : undefined;
  return response.send({ content: `Hello, ${name || 'World!'}` });
});

app.use('/api/stop', (request, response) => {
  isOnline = false;
  return response.send('Stopping HTTP server');
});

const options = {
  livenessCallback: (request, response) => {
    return isOnline ? response.send('OK') : response.sendStatus(500);
  }
};

probe(app, options);

module.exports = app;
